### PR TITLE
Permissioned restriction rollout proposal

### PR DESCRIPTION
**Pull request by y0sher**
This is my take on smooth way to rollout the permissioned feature into testnet and out of mainnet. 
this is of course based on @egorka-sereda branch and code of this feature you can see here - https://github.com/bloxapp/ssv/pull/934

@lior-blox 
@moshe-blox 
@egorka-sereda 
please take a look and tell me if something looks wrong.

### Pending
- [x] - lint
- [x] - update unit test
- [ ] - QA

### CODE REVIEW

File: network/peers/connections/handshaker.go [hunk 1]
```diff
@@ -187,22 +206,27 @@ func (h *handshaker) Handshake(logger *zap.Logger, conn libp2pnetwork.Conn) erro
 	if err := h.preHandshake(conn); err != nil {
 		return errors.Wrap(err, "could not perform pre-handshake")
 	}
-	sni, err := h.nodeInfoFromStream(logger, conn)
+
+	var ani records.AnyNodeInfo
+
+	ani, err = h.nodeInfoFromStream(logger, conn)
```
Comment (egorka-sereda) [inline network/peers/connections/handshaker.go:212]: do we need `var` initiation here?
We can just `ani, err := h.nodeInfoFromStream(logger, conn)`

File: network/records/signed_node_info.go [hunk 2]
```diff
@@ -12,12 +12,27 @@ import (
...
+type AnyNodeInfo interface {
+	Codec() []byte
+	Domain() string
+	MarshalRecord() ([]byte, error)
+	UnmarshalRecord(data []byte) error
+	Seal(netPrivateKey crypto.PrivKey) ([]byte, error)
+	Consume(data []byte) error
+	GetNodeInfo() *NodeInfo
+}
+
```
Comment (egorka-sereda) [inline network/records/signed_node_info.go:24]: we can replace 
```
        Codec() []byte
 	Domain() string
 	MarshalRecord() ([]byte, error)
 	UnmarshalRecord(data []byte) error
 ```
 
 with just 
 ```
        record.Record
 ```

Comment (egorka-sereda) [reply network/records/signed_node_info.go:24]: Could we move `AnyNodeInfo` interface to a separate file?

**Review by egorka-sereda — APPROVED** [review]
Looks awesome!
Thank you very much for your ideas!!!

File: network/peers/connections/handshaker.go (hunk 1 above)
Comment (y0sher) [reply network/peers/connections/handshaker.go:211]: you are right its not a must

File: network/peers/connections/filters.go [hunk 3]
```diff
@@ -42,7 +47,11 @@ func SenderRecipientIPsCheckFilter(me peer.ID) HandshakeFilter { // for some rea
 }
 
 func SignatureCheckFilter() HandshakeFilter {
-	return func(sender peer.ID, sni records.SignedNodeInfo) error {
+	return func(sender peer.ID, ani records.AnyNodeInfo) error {
+		sni, ok := ani.(*records.SignedNodeInfo)
+		if !ok {
+			return fmt.Errorf("wrong format nodeinfo sent")
```
[the remark this answers is not in the document]
Comment (y0sher) [reply network/peers/connections/filters.go:53]: both work you are welcome to change

File: network/peers/connections/filters.go [hunk 4]
```diff
@@ -67,24 +76,26 @@ func SignatureCheckFilter() HandshakeFilter {
 }
 
 func RegisteredOperatorsFilter(logger *zap.Logger, nodeStorage storage.Storage, keysConfigWhitelist []string) HandshakeFilter {
-	return func(sender peer.ID, sni records.SignedNodeInfo) error {
+	return func(sender peer.ID, ani records.AnyNodeInfo) error {
+		sni, ok := ani.(*records.SignedNodeInfo)
+		if !ok {
+			return fmt.Errorf("wrong format nodeinfo sent")
```
[the remark this answers is not in the document]
Comment (y0sher) [reply network/peers/connections/filters.go:82]: ditto

File: network/peers/connections/filters.go [hunk 5]
```diff
@@ -19,16 +19,21 @@ var AllowedDifference = 30 * time.Second
...
 		return nil
 	}
 }
 
 func SenderRecipientIPsCheckFilter(me peer.ID) HandshakeFilter { // for some reason we're loosing 'me' value
-	return func(sender peer.ID, sni records.SignedNodeInfo) error {
+	return func(sender peer.ID, ani records.AnyNodeInfo) error {
+		sni, ok := ani.(*records.SignedNodeInfo)
+		if !ok {
+			return fmt.Errorf("wrong format nodeinfo sent")
```
[the remark this answers is not in the document]
Comment (y0sher) [reply network/peers/connections/filters.go:35]: ditto

File: network/records/signed_node_info.go (hunk 2 above)
Comment (egorka-sereda) [reply network/records/signed_node_info.go:24]: okay, fine
I will read in my spare time :)

File: network/peers/connections/handshaker_test.go [hunk 6]
```diff
@@ -69,3 +68,130 @@ func TestHandshake(t *testing.T) {
...
+			},
+		}
+
+		require.ErrorIs(t, td.Handshaker.Handshake(logging.TestLogger(t), td.Conn), tc.expectedErr)
+	}
+}
+
+// TestHandshakePermissionedFlow tests Handshake() Permissioned flow
+func TestHandshakePermissionedFlow(t *testing.T) {
+	td := getTestingData(t)
```
[the remark this answers is not in the document]
Comment (egorka-sereda) [reply network/peers/connections/handshaker_test.go:121]: fixed

File: .k8/yamls-stage/ssv-node-10-deployment.yml [hunk 7]
```diff
@@ -134,6 +134,12 @@ spec:
           value: "152834"
         - name: BUILDER_PROPOSALS
           value: "true"
+        - name: PERMISSIONED_ACTIVE_SLOT
+          value: "5657651"
+        - name: PERMISSIONED_DEACTIVE_SLOT
+          value: "1117464639"
+        - name: WHITELISTED_KEYS
```
Comment (lior-blox) [inline .k8/yamls-stage/ssv-node-10-deployment.yml:141]: WHITELISTED_KEYS should be file and support

File: .k8/yamls-stage/ssv-node-v2-4-deployment.yml [hunk 8]
```diff
@@ -138,6 +138,12 @@ spec:
           value: 'false'
         - name: GENESIS_EPOCH
           value: "152834"
+        - name: PERMISSIONED_ACTIVE_SLOT
+          value: "5657651"
+        - name: PERMISSIONED_DEACTIVE_SLOT
+          value: "7464639"
```
Comment (lior-blox) [inline .k8/yamls-stage/ssv-node-v2-4-deployment.yml:144]: move to config map or secret

File: cli/operator/node.go [hunk 9]
```diff
@@ -98,6 +97,10 @@ var StartNodeCmd = &cobra.Command{
 		}
 
 		cfg.P2pNetworkConfig.Ctx = cmd.Context()
+		currentSlot := uint64(eth2Network.EstimatedCurrentSlot())
+		if currentSlot > cfg.P2pNetworkConfig.PermissionedActivateSlot && currentSlot < cfg.P2pNetworkConfig.PermissionedDeactivateSlot {
```
Comment (lior-blox) [inline cli/operator/node.go:101]: change to >=

File: integration/qbft/tests/regular_test.go [hunk 10]
```diff
@@ -102,7 +102,7 @@ func TestRegular10CommitteeScenario(t *testing.T) {
 func regularValidator() func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
 	return func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
 		require.Equal(t, int(qbft.FirstHeight), int(actual.State.Height), "height not matching") //int conversion needs to show correct output from require
-		require.Equal(t, int(qbft.FirstRound), int(actual.State.Round), "round not matching")
+		require.Less(t, int(actual.State.Round), 8, "round not matching")
```
[the remark this answers is not in the document]
Comment (lior-blox) [reply integration/qbft/tests/regular_test.go:105]: revert

File: network/p2p/config.go [hunk 11]
```diff
@@ -77,6 +77,13 @@ type Config struct {
 	FullNode bool
 
 	GetValidatorStats network.GetValidatorStats
+
+	PermissionedActivateSlot   uint64 `yaml:"PermissionedActivateSlot" env:"PERMISSIONED_ACTIVE_SLOT" env-description:"On which slot to start only accepting peers that are operators registered in the contract"`
```
Comment (lior-blox) [inline network/p2p/config.go:81]: add defaults max slot

File: network/p2p/config.go [hunk 12]
```diff
@@ -77,6 +77,13 @@ type Config struct {
...
+	PermissionedDeactivateSlot uint64 `yaml:"PermissionedDeactivateSlot" env:"PERMISSIONED_DEACTIVE_SLOT" env-description:"On which slot to start accepting operators all peers"`
```
Comment (lior-blox) [inline network/p2p/config.go:82]: add defaults max slot

File: network/records/signed_node_info.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+	ser := newSerializable(parts...)
+	return json.Marshal(ser)
+}
+
+func (sni *SignedNodeInfo) UnmarshalRecord(data []byte) error {
+	ser := serializable{}
+	if err := json.Unmarshal(data, &ser); err != nil {
+		return err
+	}
+
```
Comment (moshe-blox) [inline network/records/signed_node_info.go:70]: Won't the code below panic on invalid data?

I'd add..
```go
if len(ser.Entries) != 6 {
  // return an error
  }
  ```

**Review by moshe-blox — CHANGES_REQUESTED** [review]

File: cli/operator/node.go [hunk 14]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"encoding/hex"
 	"fmt"
+	"github.com/bloxapp/ssv/logging"
```
Comment (moshe-blox) [inline cli/operator/node.go:7]:
```suggestion
	"encoding/l"
	"fmt"
	"github.com/bloxapp/ssv/logging"
```

**Pull request merged**